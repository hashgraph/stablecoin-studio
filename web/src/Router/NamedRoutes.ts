export enum NamedRoutes {
	Dashboard = 'dashboard',
	Login = 'login',
	Operations = 'operations',
	Roles = 'roles',
	Details = 'details',
	StableCoinNotSelected = 'stableCoinNotSelected',
	StableCoinCreation = 'stableCoinCreation',
}
