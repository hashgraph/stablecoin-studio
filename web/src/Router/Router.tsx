import { useEffect, useState } from 'react';
import { Navigate, Route, Routes, Outlet } from 'react-router-dom';
import Layout from '../layout/Layout';
import { RoutesMappingUrl } from './RoutesMappingUrl';
import CashInOperation from '../views/Operations/CashIn';
import Dashboard from '../views/Dashboard';
import GiveRole from '../views/Roles/GiveRole';
import Login from '../views/Login';
import Operations from '../views/Operations';
import Roles from '../views/Roles';
import StableCoinCreation from '../views/StableCoinCreation/StableCoinCreation';
import StableCoinNotSelected from '../views/ErrorPage/StableCoinNotSelected';
import SDKService, { HashConnectConnectionState } from '../services/SDKService';

const PrivateRoute = ({ status }: { status?: HashConnectConnectionState }) => {
	return (
		<Layout>
			{status === HashConnectConnectionState.Paired ? (
				<Outlet />
			) : (
				<Navigate to={RoutesMappingUrl.login} replace />
			)}
		</Layout>
	);
};

const OnboardingRoute = ({ status }: { status?: HashConnectConnectionState }) => {
	return status !== HashConnectConnectionState.Paired ? (
		<Outlet />
	) : (
		<Navigate to={RoutesMappingUrl.dashboard} replace />
	);
};

const Router = () => {
	const [status, setStatus] = useState<HashConnectConnectionState>();

	useEffect(() => {
		getStatus();
	}, []);

	const getStatus = async () => {
		const status = await SDKService.getStatus();
		setStatus(status);
	};

	return (
		<main>
			<Routes>
				{/* Public routes */}
				<Route element={<OnboardingRoute status={status} />}>
					<Route path={RoutesMappingUrl.login} element={<Login />} />
				</Route>
				{/* Private routes */}
				<Route element={<PrivateRoute status={status} />}>
					<Route path={RoutesMappingUrl.cashIn} element={<CashInOperation />} />
					<Route path={RoutesMappingUrl.dashboard} element={<Dashboard />} />
					<Route path={RoutesMappingUrl.giveRole} element={<GiveRole />} />
					<Route path={RoutesMappingUrl.operations} element={<Operations />} />
					<Route path={RoutesMappingUrl.roles} element={<Roles />} />
					<Route path={RoutesMappingUrl.stableCoinCreation} element={<StableCoinCreation />} />
					<Route
						path={RoutesMappingUrl.stableCoinNotSelected}
						element={<StableCoinNotSelected />}
					/>
					<Route path='*' element={<Navigate to={RoutesMappingUrl.dashboard} />} />
				</Route>
			</Routes>
		</main>
	);
};

export default Router;
