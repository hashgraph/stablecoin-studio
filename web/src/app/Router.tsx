import React from 'react';
import { Navigate, Route, Routes } from 'react-router-dom';
import Layout from '../layout/Layout';
import Dashboard from '../views/Dashboard';
import Login from '../views/Login';

interface Props {
	handleRoute: {
		component: React.FC;
		type?: string;
	};
	roleRoute: {
		component: React.FC;
		roles: string[];
	};
}

const HandleRoute = ({ component: Component, type = 'public' }: Props['handleRoute']) => {
	return <Component />;
};

const WrapperRoutesLayout = () => {
	return (
		<div>
			<Layout>
				<Routes>
					{/* Dashboard route */}
					<Route path='/dashboard' element={<Dashboard />} />

					{/* Default routes */}
					<Route path='*' element={<Navigate to='/dashboard' />} />
				</Routes>
			</Layout>
		</div>
	);
};

const Router = () => {
	return (
		<main>
			<Routes>
				{/* Public routes */}
				<Route path='/' element={<HandleRoute component={Login} />} />
				<Route path='/login' element={<HandleRoute component={Login} />} />

				{/* Private routes */}
				<Route path='/*' element={<HandleRoute component={WrapperRoutesLayout} type='private' />} />
			</Routes>
		</main>
	);
};

export default Router;
