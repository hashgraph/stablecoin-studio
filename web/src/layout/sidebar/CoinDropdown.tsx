import { Box } from '@chakra-ui/react';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import SearchSelectController from '../../components/Form/SearchSelectController';

const CoinDropdown = () => {
	const { t } = useTranslation('global');
	const { control } = useForm();
	const styles = {
		menuList: { maxH: '220px', overflowY: 'auto', bg: 'brand.white' },
		wrapperOpened: { borderWidth: '0' },
	};

	return (
		<Box w={{ base: 'full', md: '250px' }}>
			<SearchSelectController
				control={control}
				styles={styles}
				name='coin-dropdown'
				options={[
					{
						label: 'uno',
						value: '1',
					},
					{
						label: 'dos',
						value: 2,
					},
					{
						label: 'dos',
						value: Math.random(),
					},
					{
						label: 'dos',
						value: Math.random(),
					},
					{
						label: 'dos',
						value: Math.random(),
					},
					{
						label: 'dos',
						value: Math.random(),
					},
					{
						label: 'dos',
						value: Math.random(),
					},
					{
						label: 'dos',
						value: Math.random(),
					},
					{
						label: 'dos',
						value: Math.random(),
					},
					{
						label: 'dos',
						value: Math.random(),
					},
					{
						label: 'dos',
						value: Math.random(),
					},
					{
						label: 'dos',
						value: Math.random(),
					},
					{
						label: 'dos',
						value: Math.random(),
					},
					{
						label: 'dos',
						value: Math.random(),
					},
					{
						label: 'ult',
						value: Math.random(),
					},
				]}
				placeholder={t('sidebar.coinDropdown.placeholder')}
			/>
		</Box>
	);
};

export default CoinDropdown;
