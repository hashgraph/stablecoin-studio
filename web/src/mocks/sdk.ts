import PublicKey from 'hedera-stable-coin-sdk/build/esm/src/domain/context/account/PublicKey.js';
import ContractId from 'hedera-stable-coin-sdk/build/esm/src/domain/context/contract/ContractId.js';
import BigDecimal from 'hedera-stable-coin-sdk/build/esm/src/domain/context/shared/BigDecimal.js';
import { HederaId } from 'hedera-stable-coin-sdk/build/esm/src/domain/context/shared/HederaId.js';

export const mockedStableCoinsList = {
	coins:[
		{ symbol: 'HBAR', id: '0.0.123' },
		{ symbol: 'EXP', id: '0.0.13234' },
		{ symbol: 'MCC', id: '0.0.48451290'},
 		{ symbol: 'MCC2', id: '0.0.48471242'},
 		{ symbol: 'MCC3', id: '0.0.48478857'}
	]
};

export const mockedSelectedStableCoin = {
	tokenId: HederaId.from('0.0.49319786'),
	name: 'TEST_ACCELERATOR_SC',
	symbol: 'TEST',
	decimals: 6,
	initialSupply: BigDecimal.fromString('1000.', 6),
	totalSupply: BigDecimal.fromString('1000.', 6),
	maxSupply: BigDecimal.fromString('1000', 6),
	proxyAddress: ContractId.from('0.0.49319785'),
	evmProxyAddress: '0000000000000000000000000000000002f08f69',
	treasury: HederaId.from('0.0.49319785'),
	paused: false,
	deleted: false,
	freezeDefault: false,
	autoRenewAccount: HederaId.from('0.0.49071854'),
	autoRenewAccountPeriod: 90,
	adminKey: ContractId.from('0.0.49319785'),
	kycKey: undefined,
	freezeKey: ContractId.from('0.0.49319785'),
	wipeKey: ContractId.from('0.0.49319785'),
	supplyKey: ContractId.from('0.0.49319785'),
	pauseKey: ContractId.from('0.0.49319785'),
};
export const mockedStableCoinCapabilities = {
	coin: mockedSelectedStableCoin,
	capabilities: [
		{ operation: 'Rescue', access: 1 },
		{ operation: 'Cash_in', access: 1 },
		{ operation: 'Burn', access: 1 },
		{ operation: 'Wipe', access: 1 },
		{ operation: 'Pause', access: 1 },
		{ operation: 'Unpause', access: 1 },
		{ operation: 'Freeze', access: 1 },
		{ operation: 'Unfreeze', access: 1 },
		{ operation: 'Delete', access: 1 },
		{ operation: 'Role_Management', access: 1 },
		{ operation: 'Admin Role', access: 1 },
		{ operation: 'Admin Role', access: 1 },
	],
	account: {
		id: HederaId.from('0.0.48471385' ),
		publicKey: new PublicKey({
			key: '03cce92867cd5e08c67da3fdd0bbae217f7ac73237e2d9058f7551e4ed6d9bf5ce',
			type: 'ECDSA_SECP256K1',
		}),
	},
};

export const mockedWalletData = {
  topic: 'da96975a-ebd9-4382-98f8-41fe25bc2b6b',
  pairingString: 'eWNvbiI6ImNzbHRpQWNjb3VudCI6ZmFsc2V9',
  encryptionKey: 'XdXXecXX-XcXc-XbaX-bXXd-XXXbXXXXcXXX',
  savedPairings: [
    {
      accountIds: ['0.0.48450590'],
      metadata: {
        name: 'HashPack',
        description: 'An example wallet',
        icon: 'https://www.hashpack.app/img/logo.svg',
        url: 'chrome&#45;extension&#58;&#47;&#47;gjagmgiddbbciopjhllkdnddhcglnemk',
        encryptionKey: 'XdXXecXX-XcXc-XbaX-bXXd-XXXbXXXXcXXX',
      },
      network: 'testnet',
      topic: 'da96975a-ebd9-4382-98f8-41fe25bc2b6b',
      origin: 'chrome-extension://gjagmgiddbbciopjhllkdnddhcglnemk',
      encryptionKey: 'XdXXecXX-XcXc-XbaX-bXXd-XXXbXXXXcXXX',
      lastUsed: 1664888808450,
    },
  ],
}

export const mockedFoundWallets = ['HashPack'];