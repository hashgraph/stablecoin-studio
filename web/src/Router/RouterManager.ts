import type { ReverseParams } from 'named-urls';
import { reverse } from 'named-urls';
import type { History } from 'history';
import { RoutesMappingUrl } from './RoutesMappingUrl';
import { NamedRoutes } from './NamedRoutes';

export class BaseRouterManager {
	constructor(private routes: Record<NamedRoutes, string> = RoutesMappingUrl) {}

	to(
		history: History,
		namedUrl: NamedRoutes,
		params?: ReverseParams,
		state?: object,
		extra?: string,
	) {
		return history.push(reverse(`${this.routes[namedUrl]}${extra || ''}`, params), state);
	}

	getUrl(namedUrl: NamedRoutes, params?: ReverseParams, extra?: string) {
		return reverse(`${this.routes[namedUrl]}${extra || ''}`, params);
	}
}

export const RouterManager = new BaseRouterManager(RoutesMappingUrl);
