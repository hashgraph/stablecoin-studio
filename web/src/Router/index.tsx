import { Box, Heading } from '@chakra-ui/react';
import { useTranslation } from 'react-i18next';
import BaseContainer from '../components/BaseContainer';
import { NamedRoutes } from './NamedRoutes';
import GridDirectAccess from '../components/GridDirectAccess';
import { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import {
	SELECTED_WALLET_CAPABILITIES,
	SELECTED_WALLET_COIN,
	SELECTED_WALLET_PAIRED_ACCOUNTID,
} from '../store/slices/walletSlice';
import type { DirectAccessProps } from '../components/DirectAccess';
import type { IAccountToken } from '../interfaces/IAccountToken';
import type { IExternalToken } from '../interfaces/IExternalToken';
// import type { AppDispatch } from '../../store/store.js';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { useRefreshCoinInfo } from '../hooks/useRefreshCoinInfo';
import { Access, Operation, StableCoinRole } from 'hedera-stable-coin-sdk';

const Operations = () => {
	const { t } = useTranslation('operations');

	const selectedStableCoin = useSelector(SELECTED_WALLET_COIN);
	const accountId = useSelector(SELECTED_WALLET_PAIRED_ACCOUNTID);
	const capabilities = useSelector(
		SELECTED_WALLET_CAPABILITIES,
	);

	const [disabledFeatures, setDisabledFeatures] = useState({
		cashIn: false,
		burn: false,
		balance: false,
		rescue: false,
		wipe: false,
		freeze: false,
		pause: false,
		delete: false,
	});

	useRefreshCoinInfo();

	useEffect(() => {
		if (selectedStableCoin) {
			getAvailableFeatures();
		}
	}, [selectedStableCoin, capabilities]);

	const getAvailableFeatures = () => {
		let isExternalToken = false;
		let roles = [];
		const tokensAccount = localStorage?.tokensAccount;
		if (tokensAccount) {
			const tokensAccountParsed = JSON.parse(tokensAccount);
			if (tokensAccountParsed) {
				const myAccount = tokensAccountParsed.find(
					(acc: IAccountToken) => acc.id === accountId?.toString(),
				);
				if (myAccount) {
					const externalToken = myAccount?.externalTokens.find(
						(coin: IExternalToken) => coin.id === selectedStableCoin?.tokenId?.toString(),
					);
					if (externalToken) {
						isExternalToken = true;
						roles = externalToken.roles.map((role: string) => role);
					}
				}
			}
		}

		function getAccessByOperation(operation: Operation): Access | undefined {
			return capabilities?.capabilities.filter((capability) => {
				return (capability.operation === operation);
			})[0].access ?? undefined;
		}

		const operations = capabilities?.capabilities.map((x) => x.operation);
		const areDisabled = {
			cashIn: !isExternalToken
				? !operations?.includes(Operation.CASH_IN)
				: (!operations?.includes(Operation.CASH_IN) ||
				  (operations?.includes(Operation.CASH_IN) && ((getAccessByOperation(Operation.CASH_IN) !== Access.HTS) && 
				  !roles.includes(StableCoinRole.CASHIN_ROLE)))),
			burn: !isExternalToken
				? !operations?.includes(Operation.BURN)
				: (!operations?.includes(Operation.BURN) ||
				(operations?.includes(Operation.BURN) && ((getAccessByOperation(Operation.BURN) !== Access.HTS) &&
				!roles.includes(StableCoinRole.BURN_ROLE)))),
			balance: false,
			rescue: !isExternalToken
				? !operations?.includes(Operation.RESCUE)
				: (!operations?.includes(Operation.RESCUE) ||
				(operations?.includes(Operation.RESCUE) && ((getAccessByOperation(Operation.RESCUE) !== Access.HTS) && 
				!roles.includes(StableCoinRole.RESCUE_ROLE)))),
			wipe: !isExternalToken
				? !operations?.includes(Operation.WIPE)
				: (!operations?.includes(Operation.WIPE) ||
				(operations?.includes(Operation.WIPE) && ((getAccessByOperation(Operation.WIPE) !== Access.HTS) && 
				!roles.includes(StableCoinRole.WIPE_ROLE)))),
			freeze: !isExternalToken
				? !operations?.includes(Operation.FREEZE)
				: (!operations?.includes(Operation.FREEZE) ||
				(operations?.includes(Operation.FREEZE) && ((getAccessByOperation(Operation.FREEZE) !== Access.HTS) &&
				!roles.includes(StableCoinRole.FREEZE_ROLE)))),
			pause: !isExternalToken
				? !operations?.includes(Operation.PAUSE)
				: (!operations?.includes(Operation.PAUSE) ||
				(operations?.includes(Operation.PAUSE) && ((getAccessByOperation(Operation.PAUSE) !== Access.HTS) &&
				!roles.includes(StableCoinRole.PAUSE_ROLE)))),
			delete: !isExternalToken
				? !operations?.includes(Operation.DELETE)
				: (!operations?.includes(Operation.DELETE) ||
				(operations?.includes(Operation.DELETE) && ((getAccessByOperation(Operation.DELETE) !== Access.HTS) && 
				!roles.includes(StableCoinRole.DELETE_ROLE)))),
								
		};
		setDisabledFeatures(areDisabled);
	};

	const directAccesses: DirectAccessProps[] = [
		{
			icon: 'ArrowDown',
			route: NamedRoutes.CashIn,
			title: t('cashInOperation'),
			isDisabled: disabledFeatures?.cashIn,
		},
		{
			icon: 'ArrowUp',
			route: NamedRoutes.Burn,
			title: t('burnOperation'),
			isDisabled: disabledFeatures?.burn,
		},
		{
			icon: 'Coin',
			route: NamedRoutes.Balance,
			title: t('getBalanceOperation'),
			isDisabled: disabledFeatures?.balance,
		},
		{
			icon: 'ArrowsDownUp',
			route: NamedRoutes.RescueTokens,
			title: t('rescueOperation'),
			isDisabled: disabledFeatures?.rescue,
		},
		{
			icon: 'MinusCircle',
			route: NamedRoutes.Wipe,
			title: t('wipeOperation'),
			isDisabled: disabledFeatures?.wipe,
		},
		{
			icon: 'Snowflake',
			route: NamedRoutes.Freeze,
			title: t('freezeOperation'),
			isDisabled: disabledFeatures?.freeze,
		},
		{
			icon: 'SunHorizon',
			route: NamedRoutes.Unfreeze,
			title: t('unfreezeOperation'),
			isDisabled: disabledFeatures?.freeze,
		},
		{
			icon: 'Warning',
			route: NamedRoutes.DangerZone,
			title: t('dangerZoneOperation'),
			isDisabled: disabledFeatures?.pause && disabledFeatures.delete,
		},
	];

	return (
		<BaseContainer title={t('title')}>
			<Box p={{ base: 4, md: '128px' }}>
				<Heading fontSize='20px' fontWeight='600' mb={14} data-testid='subtitle'>
					{t('subtitle')}
				</Heading>
				<GridDirectAccess directAccesses={directAccesses} />
			</Box>
		</BaseContainer>
	);
};

export default Operations;
