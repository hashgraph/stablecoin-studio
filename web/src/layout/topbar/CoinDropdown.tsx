import { Box } from '@chakra-ui/react';
import { useEffect, useState } from 'react';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import SearchSelectController from '../../components/Form/SearchSelectController';
import type { Option } from '../../components/Form/SelectController';
import SDKService from '../../services/SDKService';
import type { AppDispatch } from '../../store/store';
import {
	getStableCoinList,
	SELECTED_WALLET_COIN,
	SELECTED_WALLET_PAIRED_ACCOUNTID,
	STABLE_COIN_LIST,
	walletActions,
} from '../../store/slices/walletSlice';
import { RouterManager } from '../../Router/RouterManager';
import { matchPath, useLocation, useNavigate } from 'react-router-dom';
import { NamedRoutes } from '../../Router/NamedRoutes';
import { HashPackAccount } from 'hedera-stable-coin-sdk';

const CoinDropdown = () => {
	const stableCoinList = useSelector(STABLE_COIN_LIST);
	const selectedStableCoin = useSelector(SELECTED_WALLET_COIN);
	const accountId = useSelector(SELECTED_WALLET_PAIRED_ACCOUNTID);
	const dispatch = useDispatch<AppDispatch>();
	const [options, setOptions] = useState<Option[]>([]);
	const navigate = useNavigate();
	const location = useLocation();
	const isInStableCoinNotSelected = !!matchPath(
		location.pathname,
		RouterManager.getUrl(NamedRoutes.StableCoinNotSelected),
	);

	useEffect(() => {
		if (selectedStableCoin && isInStableCoinNotSelected) {
			RouterManager.to(navigate, NamedRoutes.Operations);
		}
	}, [selectedStableCoin]);

	useEffect(() => {
		if (accountId) {
			const id = new HashPackAccount(accountId);
			dispatch(getStableCoinList(id));
		}
	}, [accountId]);

	useEffect(() => {
		formatOptionsStableCoins();
	}, [stableCoinList]);

	const formatOptionsStableCoins = async () => {
		if (stableCoinList) {
			const options = stableCoinList.map(({ id, symbol }) => ({
				label: `${id} - ${symbol}`,
				value: id,
			}));
			setOptions(options);
		}
	};

	const formatSupplyParam = ({
		supply,
		decimals,
	}: {
		supply: string | undefined;
		decimals: number;
	}) => {
		if (supply !== '0') return supply?.slice(0, -decimals);

		return supply;
	};

	const handleSelectCoin = async (event: any) => {
		const selectedCoin = event.value;

		const stableCoinDetails = await SDKService.getStableCoinDetails({
			id: selectedCoin,
		});

		dispatch(
			walletActions.setSelectedStableCoin({
				tokenId: stableCoinDetails?.tokenId,
				initialSupply: formatSupplyParam({
					supply: stableCoinDetails?.initialSupply?.toString(),
					decimals: stableCoinDetails?.decimals ?? 0,
				}),
				totalSupply: formatSupplyParam({
					supply: stableCoinDetails?.totalSupply?.toString(),
					decimals: stableCoinDetails?.decimals ?? 0,
				}),
				maxSupply: formatSupplyParam({
					supply: stableCoinDetails?.maxSupply?.toString(),
					decimals: stableCoinDetails?.decimals ?? 0,
				}),
				name: stableCoinDetails?.name,
				symbol: stableCoinDetails?.symbol,
				decimals: stableCoinDetails?.decimals,
				id: stableCoinDetails?.tokenId,
				treasuryId: stableCoinDetails?.treasuryId,
				memo: stableCoinDetails?.memo,
				adminKey:
					stableCoinDetails?.adminKey && JSON.parse(JSON.stringify(stableCoinDetails.adminKey)),
				kycKey: stableCoinDetails?.kycKey && JSON.parse(JSON.stringify(stableCoinDetails.kycKey)),
				freezeKey:
					stableCoinDetails?.freezeKey && JSON.parse(JSON.stringify(stableCoinDetails.freezeKey)),
				wipeKey:
					stableCoinDetails?.wipeKey && JSON.parse(JSON.stringify(stableCoinDetails.wipeKey)),
				supplyKey:
					stableCoinDetails?.supplyKey && JSON.parse(JSON.stringify(stableCoinDetails.supplyKey)),
			}),
		);
	};

	const { t } = useTranslation('global');
	const { control } = useForm();
	const styles = {
		menuList: {
			maxH: '244px',
			overflowY: 'auto',
			bg: 'brand.white',
			boxShadow: 'down-black',
			p: 4,
		},
		wrapperOpened: { borderWidth: '0' },
	};

	return (
		<Box w={{ base: 'full', md: '280px' }} data-testid='coin-dropdown'>
			<SearchSelectController
				control={control}
				styles={styles}
				name='coin-dropdown'
				options={options}
				placeholder={t('topbar.coinDropdown.placeholder')}
				iconStyles={{ color: 'brand.primary200' }}
				onChangeAux={handleSelectCoin}
				noOptionsMessage={() => t('topbar.coinDropdown.noStableCoin')}
			/>
		</Box>
	);
};

export default CoinDropdown;
