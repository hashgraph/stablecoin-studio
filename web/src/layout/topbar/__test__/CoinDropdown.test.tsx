import CoinDropdown from '../CoinDropdown';
import { render } from '../../../test/';
import { waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { getStableCoinsFull } from '../../../mocks/sdk';
import { Provider } from 'react-redux';
import type { MockStoreEnhanced } from 'redux-mock-store';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { RouterManager } from '../../../Router/RouterManager';
import { NamedRoutes } from '../../../Router/NamedRoutes';

jest.mock('react-select', () =>
	// eslint-disable-next-line react/display-name
	({ name, options, value, onChange, onBlurAux, placeholder }: any) => {
		function handleChange(event: any) {
			const option = options.find((option: any) => option.value === event.currentTarget.value);

			onChange(option);
		}

		return (
			<>
				<span data-testid='select-placeholder'>{placeholder} </span>
				<select
					data-testid={`select-${name}`}
					value={value}
					onChange={handleChange}
					onBlur={onBlurAux}
				>
					{options.map(({ label, value }: any) => (
						<option key={value} value={value}>
							{label}
						</option>
					))}
				</select>
			</>
		);
	},
);
jest.mock('../../../Router/RouterManager', () => ({
	RouterManager: {
		to: jest.fn(),
	},
}));
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const defaultStore = mockStore({
	wallet: {
		stableCoinList: getStableCoinsFull,
	},
});

let localStore: MockStoreEnhanced<unknown, {}>;
describe(`<${CoinDropdown.name} />`, () => {
	const factoryComponent = ({
		store = defaultStore,
	}: {
		store?: MockStoreEnhanced<unknown, {}>;
	} = {}) => {
		localStore = store;
		return render(
			<Provider store={localStore}>
				<CoinDropdown />
			</Provider>,
		);
	};

	test('should render correctly', () => {
		const component = factoryComponent();

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should show all stable coins', async () => {
		const component = factoryComponent();

		const select = component.getByTestId('select-coin-dropdown');
		userEvent.click(select);
		const coinLabel = `${getStableCoinsFull[0].id} - ${getStableCoinsFull[0].symbol}`;

		await waitFor(() => {
			expect(component.getByText(coinLabel)).toBeInTheDocument();
		});
	});

	test('should be able to choose one coin', async () => {
		const component = factoryComponent();

		const select = component.getByTestId('select-coin-dropdown');
		userEvent.click(select);

		await waitFor(() => {
			const coinLabel = `${getStableCoinsFull[0].id} - ${getStableCoinsFull[0].symbol}`;
			const option = component.getByText(coinLabel);

			userEvent.click(option);
			expect((select as HTMLInputElement).value).toEqual(getStableCoinsFull[0].id);
		});
		// TODO: validate onchange fn
	});

	test('should go to StableCoinNotSelected if coin isnt selected ', async () => {
		const store = mockStore({
			wallet: {},
		});
		const anything = expect.any(Function);

		factoryComponent({ store });

		await waitFor(() => {
			expect(RouterManager.to).toHaveBeenCalledWith(anything, NamedRoutes.StableCoinNotSelected);
		});
	});
});
