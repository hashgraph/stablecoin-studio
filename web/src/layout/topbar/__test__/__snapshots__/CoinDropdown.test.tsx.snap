// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CoinDropdown /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-d9s5ly"
    data-testid="coin-dropdown"
  >
    <div
      class="chakra-stack css-1h0sumw"
    >
      <div
        class="chakra-form-control css-1kxonj9"
        data-testid="form_control"
        role="group"
      >
        <div
          class="css-frrdvf"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class=" css-15j38w8"
          >
            <div
              class="css-3y2vuh"
            >
              <span
                class="chakra-text css-4pvr19"
                data-testid="select-placeholder"
              >
                Choose Stable coin
              </span>
              <div
                class=" css-svnjlw"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-2-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=" css-146k1ar-input"
                  id="field-:r0:"
                  role="combobox"
                  spellcheck="false"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class=" css-hfbj6y"
            >
              <div
                class="css-1aabyhj"
              >
                <svg
                  class="chakra-icon css-zsadqz"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 256 256"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    fill="none"
                    height="256"
                    width="256"
                  />
                  <polyline
                    fill="none"
                    points="208 96 128 176 48 96"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                  />
                </svg>
              </div>
            </div>
          </div>
          <input
            name="coin-dropdown"
            type="hidden"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
